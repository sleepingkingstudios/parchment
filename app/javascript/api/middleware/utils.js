import { convertToArray } from '../../utils/array';

const mapProperty = (ary, propName) => {
  if (typeof propName === 'undefined' || propName === null) { return ary; }

  return ary.map(item => item[propName]);
};

const injectMiddlewareAfter = (ary, middleware, target) => {
  const copy = ary.slice();

  if (target === ':all') {
    copy.splice(target.length, 0, middleware);

    return copy;
  }

  const index = ary.findIndex(item => (item.type === target));

  if (index === -1) {
    throw new Error(`Unable to inject middleware after "${target}"`);
  }

  copy.splice(index + 1, 0, middleware);

  return copy;
};

const injectMiddlewareBefore = (ary, middleware, target) => {
  const copy = ary.slice();

  if (target === ':all') {
    copy.splice(0, 0, middleware);

    return copy;
  }

  const index = ary.findIndex(item => (item.type === target));

  if (index === -1) {
    throw new Error(`Unable to inject middleware before "${target}"`);
  }

  copy.splice(index, 0, middleware);

  return copy;
};

const applyDirective = (ary, directive) => {
  const { middleware } = directive;

  if (directive.after) {
    return injectMiddlewareAfter(ary, middleware, directive.after);
  }

  if (directive.before) {
    return injectMiddlewareBefore(ary, middleware, directive.before);
  }

  return ary;
};

export const applyMiddleware = (original, middleware) => {
  if (middleware.length === 0) { return original; }

  let fn = original;

  middleware.reverse().forEach((middlewareFn) => { fn = middlewareFn(fn); });

  return fn;
};

export const injectMiddleware = (middleware, directives) => (
  directives.reduce(applyDirective, middleware)
);

export const selectMiddleware = (maybeMiddleware, propName) => {
  const middleware = convertToArray(maybeMiddleware);
  const selected = [];

  mapProperty(middleware, propName).forEach((item) => {
    if (typeof item === 'function') { selected.push(item); }
  });

  return selected;
};

export const wrapMiddleware = (middleware, next) => {
  if (typeof middleware !== 'function') { return next; }

  return middleware(next);
};
