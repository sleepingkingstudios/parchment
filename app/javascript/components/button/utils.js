export const buttonStyleClass = ({ buttonStyle, outline }) => {
  switch (buttonStyle) {
    case 'link':
      return 'btn-link';
    case 'primary':
    case 'secondary':
    case 'success':
    case 'danger':
    case 'warning':
    case 'info':
    case 'light':
    case 'dark':
      return `btn-${outline ? 'outline-' : ''}${buttonStyle}`;
    default:
      return `btn-${outline ? 'outline-' : ''}primary`;
  }
};

export const buttonClass = ({
  block,
  buttonStyle,
  className,
  outline,
}) => {
  const classes = ['btn', buttonStyleClass({ buttonStyle, outline })];

  if (block) { classes.push('btn-block'); }

  if (className) { classes.push(className); }

  return classes.join(' ');
};

export const disabledOnClick = ({ preventDefault }) => { preventDefault(); };
