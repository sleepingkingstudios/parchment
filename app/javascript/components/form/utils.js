import {
  convertToArray,
} from '../../utils/array';
import {
  camelize,
  underscore,
} from '../../utils/string';

export const formatErrors = (errors) => {
  if (typeof errors === 'undefined') { return {}; }
  if (errors === null) { return {}; }

  const formatted = {};

  errors.forEach((tuple) => {
    const property = camelize(tuple[0]);
    const message = tuple[1];

    if (!formatted[property]) { formatted[property] = []; }

    formatted[property].push(message);
  });

  return formatted;
};

export const generateFieldId = ({ inputId, path, prop }) => {
  if (inputId) { return inputId; }

  const segments = convertToArray(path).slice();

  segments.push(prop);

  segments.push('input');

  return segments
    .map(str => underscore(str).replace(/_/g, '-'))
    .join('-');
};

export const getInputValue = ({ checked, type, value }) => {
  if (type === 'checkbox') { return checked; }

  return value;
};

export const coerceInputValue = (handler, fn) => (
  (event) => {
    const { target } = event;
    const value = fn(getInputValue(target));

    return handler({ target: { value } });
  }
);
