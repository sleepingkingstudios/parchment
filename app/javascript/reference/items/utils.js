import {
  capitalize,
  underscore,
} from 'utils/string';

// eslint-disable-next-line import/prefer-default-export
export const formatType = (str) => {
  if (typeof str === 'undefined' || str == null) { return ''; }

  const segments = str.split('::');

  return underscore(segments[segments.length - 1])
    .split(/[_ ]/)
    .map(capitalize)
    .join(' ');
};
