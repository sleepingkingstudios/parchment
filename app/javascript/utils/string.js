const articles = ['of', 'on', 'the'];

const camelizeMatch = (match, p1, p2) => {
  if (p2) { return p2.toUpperCase(); }

  return p1.toLowerCase();
};

export const camelize = (str) => {
  if (typeof str === 'undefined' || str == null) { return ''; }

  return str.replace(/^([A-Z])|[\s-_]+(\w)/g, camelizeMatch);
};

export const capitalize = (word) => {
  if (typeof word === 'undefined' || word == null) { return ''; }

  return word.charAt(0).toUpperCase() + word.slice(1).toLowerCase();
};

export const interpolate = (str, rxp, params) => {
  if (typeof str === 'undefined' || str == null) { return ''; }

  if (typeof rxp === 'undefined' || rxp == null) { return str; }

  return str.replace(rxp, (_, key) => params[key]);
};

export const safeCapitalize = (word) => {
  if (typeof word === 'undefined' || word == null) { return ''; }

  return word.charAt(0).toUpperCase() + word.slice(1);
};

export const slugify = (str) => {
  if (typeof str === 'undefined' || str == null) { return ''; }

  if (str.length === 0) { return ''; }

  return str
    .trim()
    .toLowerCase()
    .split(/[\s\-_]+/)
    .map(word => word.replace(/[^a-z0-9]/g, ''))
    .filter((word) => {
      if (word.length === 0) { return false; }

      return !articles.includes(word);
    })
    .join('-');
};

export const truncate = (str, len) => {
  if (typeof str === 'undefined' || str == null) { return ''; }

  if (str.length <= len) { return str; }

  return `${str.slice(0, len - 3)}...`;
};

export const underscore = (str) => {
  if (typeof str === 'undefined' || str == null) { return ''; }

  return str
    .replace(/([a-z\d])([A-Z])/g, '$1_$2')
    .replace(/([A-Z]+)([A-Z][a-z\d]+)/g, '$1_$2')
    .toLowerCase();
};

export const upperCamelize = (str) => {
  const camel = camelize(str);

  return camel.charAt(0).toUpperCase() + camel.slice(1);
};

export const titleize = (str) => {
  if (typeof str === 'undefined' || str == null) { return ''; }

  return underscore(str).split(/[\s_]/).map(capitalize).join(' ');
};
