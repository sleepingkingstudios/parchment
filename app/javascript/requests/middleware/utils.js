import { convertToArray } from '../../utils/array';

const mapProperty = (ary, propName) => {
  if (typeof propName === 'undefined' || propName === null) { return ary; }

  return ary.map(item => item[propName]);
};

export const applyMiddleware = (original, middleware) => {
  if (middleware.length === 0) { return original; }

  let fn = original;

  middleware.reverse().forEach((middlewareFn) => { fn = middlewareFn(fn); });

  return fn;
};

export const selectMiddleware = (maybeMiddleware, propName) => {
  const middleware = convertToArray(maybeMiddleware);
  const selected = [];

  mapProperty(middleware, propName).forEach((item) => {
    if (typeof item === 'function') { selected.push(item); }
  });

  return selected;
};
