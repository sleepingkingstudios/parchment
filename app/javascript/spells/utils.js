import { capitalize } from '../utils/string';
import { shortOrdinal } from '../utils/number';

export const formatComponents = (spell, includeMaterials = false) => {
  const components = [];

  if (spell.verbalComponent) { components.push('V'); }

  if (spell.somaticComponent) { components.push('S'); }

  if ((spell.materialComponent || 0).length > 0) {
    if (includeMaterials) {
      components.push(`M (${spell.materialComponent})`);
    } else {
      components.push('M');
    }
  }

  if (components.length === 0) { return 'None'; }

  return components.join(', ');
};

export const formatSchoolAndLevel = ({ level, ritual, school }) => {
  const fragments = [];

  if (level === 0) {
    fragments.push(capitalize(school));
    fragments.push('cantrip');
  } else {
    fragments.push(`${shortOrdinal(level)}-level`);
    fragments.push(school.toLowerCase());
  }

  if (ritual) { fragments.push('(ritual)'); }

  return fragments.join(' ');
};
