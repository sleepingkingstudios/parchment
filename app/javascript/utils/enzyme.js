import { valueOrDefault } from './object';
import { safeCapitalize } from './string';

export const extractInjectedComponent = (component) => {
  const injected = component.shallow().find('ComponentWithInjectedProps');

  return injected.exists() ? injected : component;
};

export const toHaveFormControl = expect => (received, expectedControl, withProps) => {
  const formControl = received.find(expectedControl);
  const expectedProps = valueOrDefault(withProps, {});

  expect(formControl).toExist();
  expect(formControl).toHaveProp(expectedProps);

  return { pass: true, message: () => '' };
};

export const toHaveFormGroup = expect => (received, expectedGroup, fn) => {
  const formGroup = received.find(expectedGroup);

  expect(formGroup).toExist();

  fn(extractInjectedComponent(formGroup).shallow());

  return { pass: true, message: () => '' };
};

export const toHaveFormInput = expect => (received, props) => {
  const {
    form,
    inputType,
    prop,
  } = props;

  expect(received).toHaveProp({ prop });

  const formInput = received.find(`${safeCapitalize(prop)}Input`);

  expect(formInput).toExist();
  expect(formInput).toHaveProp({ form });

  const shallowInput = formInput.shallow();

  expect(shallowInput).toExist();
  expect(shallowInput).toHaveDisplayName(valueOrDefault(inputType, 'FormInput'));

  return { pass: true, message: () => '' };
};
