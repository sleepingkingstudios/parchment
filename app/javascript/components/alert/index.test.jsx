import React from 'react';
import { shallow } from 'enzyme';

import Alert from './index';

import { capitalize } from '../../utils/string';

describe('<Alert />', () => {
  const defaultProps = {
    message:
      'This is a test of the Emergency Broadcast System. This is only a test.',
  };

  it('should wrap the contents in a div', () => {
    const rendered = shallow(<Alert {...defaultProps} />);

    expect(rendered).toHaveDisplayName('div');
    expect(rendered).toHaveClassName('alert');
    expect(rendered).toHaveClassName('alert-primary');
  });

  it('should render the message', () => {
    const rendered = shallow(<Alert {...defaultProps} />);
    const alertStyle = 'primary';
    const { message } = defaultProps;

    expect(rendered).toContainReact(
      <strong>{ `${capitalize(alertStyle)}:` }</strong>,
    );
    expect(rendered).toIncludeText(message);
  });

  it('should match the snapshot', () => {
    const rendered = shallow(<Alert {...defaultProps} />);

    expect(rendered).toMatchSnapshot();
  });

  describe('with alertStyle: danger', () => {
    const props = { ...defaultProps, alertStyle: 'danger' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName(`alert-${alertStyle}`);
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with alertStyle: dark', () => {
    const props = { ...defaultProps, alertStyle: 'dark' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName(`alert-${alertStyle}`);
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with alertStyle: info', () => {
    const props = { ...defaultProps, alertStyle: 'info' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName(`alert-${alertStyle}`);
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with alertStyle: light', () => {
    const props = { ...defaultProps, alertStyle: 'light' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName(`alert-${alertStyle}`);
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with alertStyle: primary', () => {
    const props = { ...defaultProps, alertStyle: 'primary' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName(`alert-${alertStyle}`);
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with alertStyle: secondary', () => {
    const props = { ...defaultProps, alertStyle: 'secondary' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName(`alert-${alertStyle}`);
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with alertStyle: success', () => {
    const props = { ...defaultProps, alertStyle: 'success' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName(`alert-${alertStyle}`);
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with alertStyle: warning', () => {
    const props = { ...defaultProps, alertStyle: 'warning' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName(`alert-${alertStyle}`);
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with alertStyle: unknown value', () => {
    const props = { ...defaultProps, alertStyle: 'enigmatic' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName('alert-primary');
    });

    it('should render the message', () => {
      const rendered = shallow(<Alert {...props} />);
      const { alertStyle, message } = props;

      expect(rendered).toContainReact(
        <strong>{ `${capitalize(alertStyle)}:` }</strong>,
      );
      expect(rendered).toIncludeText(message);
    });
  });

  describe('with className: value', () => {
    const props = { ...defaultProps, className: 'custom-alert' };

    it('should wrap the contents in a div', () => {
      const rendered = shallow(<Alert {...props} />);
      const { className } = props;

      expect(rendered).toHaveDisplayName('div');
      expect(rendered).toHaveClassName('alert');
      expect(rendered).toHaveClassName('alert-primary');
      expect(rendered).toHaveClassName(className);
    });
  });

  describe('with children: value', () => {
    const children = (
      <dl>
        <dt>What lies beyond the furthest reaches of the sky?</dt>
        <dd>
          {
            "That which will lead the lost child back to her mother's arms. "
            + 'Exile.'
          }
        </dd>
      </dl>
    );
    const props = { ...defaultProps, children };

    it('should render the children', () => {
      const rendered = shallow(<Alert {...props} />);

      expect(rendered).toContainReact(children);
      expect(rendered).not.toIncludeText('Primary:');
    });

    it('should match the snapshot', () => {
      const rendered = shallow(<Alert {...props} />);

      expect(rendered).toMatchSnapshot();
    });
  });
});
