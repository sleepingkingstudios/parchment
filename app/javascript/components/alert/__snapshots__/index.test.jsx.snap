// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> should match the snapshot 1`] = `
<div
  className="alert alert-primary"
>
  <strong>
    Primary
    :
  </strong>
   This is a test of the Emergency Broadcast System. This is only a test.
</div>
`;

exports[`<Alert /> with children: value should match the snapshot 1`] = `
<div
  className="alert alert-primary"
>
  <dl>
    <dt>
      What lies beyond the furthest reaches of the sky?
    </dt>
    <dd>
      That which will lead the lost child back to her mother's arms. Exile.
    </dd>
  </dl>
</div>
`;
